src/main/cpp/src/decimal_utils.cu

```diff
@@ -520,6 +520,111 @@ inline __device__ int precision10(chunked256 value) {
     return -1;
 }
 
+// Perform a 256-bit add in 64-bit chunks
+__device__ chunked256 add(chunked256 const &a, chunked256 const &b) {
+  chunked256 r;
+  __uint128_t add;
+  uint64_t carry = 0;
+  for (int idx = 0; idx < 4; ++idx) {
+      add = static_cast<__uint128_t>(a[idx]) + b[idx] + carry;
+      r[idx] = static_cast<uint64_t>(add);
+      carry = static_cast<uint64_t>(add >> 64);
+  }
+  return r;
+}
+
+// Functor to add two DECIMAL128 columns with rounding and overflow detection.
+struct dec128_add: public thrust::unary_function<cudf::size_type, __int128_t> {
+  dec128_add(bool *overflows, cudf::mutable_column_view const &product_view,
+                    cudf::column_view const &a_col, cudf::column_view const &b_col)
+      : overflows(overflows), a_data(a_col.data<__int128_t>()), b_data(b_col.data<__int128_t>()),
+        add_data(product_view.data<__int128_t>()),
+        a_scale(a_col.type().scale()), b_scale(b_col.type().scale()),
+        sum_scale(product_view.type().scale()) {}
+
+  __device__ __int128_t operator()(cudf::size_type const i) const {
+    chunked256 const a(a_data[i]);
+    chunked256 const b(b_data[i]);
+
+    chunked256 working_a(a_data[i]);
+    chunked256 working_b(b_data[i]);
+
+    /*
+    * The way Spark 3.4 does add is first it rescales the original numbers
```

Comment:
I am confused by this. The Spark/Java code does not appear to do what you are claiming, and your own code is making assumptions that are not true.

https://github.com/apache/spark/blob/25133684ef9991d0b0414d97ac64c70e878b4d4d/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala#L461-L472

Spark has a special case if the value is stored as a long and the scales are the same, but in all other cases it becomes a Java BigDecimal add, which according to the docs will output a scale that is the max of LHS.scale() and RHS.scale().

After that the result is rounded to the final scale of the answer.

The current code assumes that LHS.scale() is the same as RHS.scale(), which is not guaranteed to be true, and will only adjust the scales if the LHS.scale() != the output scale.

We should follow what java does.

```
int intermediate_scale = max(a_scale, b_scale);
if (a_scale != intermediate_scale) {
  a.set_scale_and_round(intermediate_scale); // you can use the code on lines 560 to 570 for this
}

if (b_scale != intermediate_scale) {
  b.set_scale_and_round(intermediate_scale);
}

// Or just use a directly and return it.
chunked256 sum = a;
sum.add(b);
if (sum_scale != intermediate_scale) {
  sum.set_scale_and_round(sum_scale);
}

overflows[i] = !sum.fits_in_128_bits();
add_data[i] = sum.as_128_bits();
```

We just need to be sure that we test this all out with negative scale decimal values.

Reply:
So, here is what I see from REPL 

```
scala> val df = Seq(("-9191008513307131620269245301.1615457290", "-9447850332473678680446404122.5624623187")).toDF("a", "b")
df: org.apache.spark.sql.DataFrame = [a: string, b: string]

scala> df.selectExpr("cast(a as decimal(38,10)) + cast(b as decimal(38,10))").show(truncate=false) 
+-------------------------------------------------------+
|(CAST(a AS DECIMAL(38,10)) + CAST(b AS DECIMAL(38,10)))|
+-------------------------------------------------------+
|-18638858845780810300715649423.724008048               |
+-------------------------------------------------------+


scala> val left = BigDecimal("-9191008513307131620269245301.1615457290") 
left: scala.math.BigDecimal = -9191008513307131620269245301.1615457290

scala> val right = BigDecimal ("-9447850332473678680446404122.5624623187")
right: scala.math.BigDecimal = -9447850332473678680446404122.5624623187

scala> left + right 
res6: scala.math.BigDecimal = -18638858845780810300715649423.7240080477
```

Why is there a discrepancy in the sum when doing it in a dataframe vs BigDecimal?

Reply:
The difference is that Spark will round it to the desired scale when it is done. That is the 

```
if (sum_scale != intermediate_scale) {
  sum.set_scale_and_round(sum_scale);
}
```
in the pseudo code I added.

In Spark it does it by calling `setScale`, with half up rounding.

```
scala>  val left = BigDecimal("-9191008513307131620269245301.1615457290") 
left: scala.math.BigDecimal = -9191008513307131620269245301.1615457290

scala>  val right = BigDecimal ("-9447850332473678680446404122.5624623187")
right: scala.math.BigDecimal = -9447850332473678680446404122.5624623187

scala> val tmp = left + right
tmp: scala.math.BigDecimal = -18638858845780810300715649423.7240080477

scala> tmp.setScale(9, BigDecimal.RoundingMode.HALF_UP)
res3: scala.math.BigDecimal = -18638858845780810300715649423.724008048
```